Add the `force_latest_compatible_version` input, and add the "auto-detect Dependabot/CompatHelper" functionality

DilumAluthge opened the pull request:
This should not be merged until https://github.com/JuliaLang/Pkg.jl/pull/2439 is merged.

christopher-dG commented on action.yml:
```diff
@@ -33,5 +36,5 @@ runs:
         # packages via `Pkg.test`.
         JULIA_PKG_SERVER: ""
-    - run: julia --color=yes --check-bounds=yes --inline=${{ inputs.inline }} --project -e 'using Pkg; Pkg.test(coverage=${{ inputs.coverage }})'
+    - run: julia --color=yes --check-bounds=yes --inline=${{ inputs.inline }} --project -e 'using Pkg; kwargs = Dict{Symbol, Any}(); kwargs[:coverage] = ${{ inputs.coverage }; ${{ inputs.force_latest_compat }} && (kwargs[:force_latest_compat] = ${{ inputs.force_latest_compat }}); Pkg.test(; kwargs...)'
```
This should probably be put on multiple lines

DilumAluthge replied:
Sure! I'm not sure what the syntax for that is.

DilumAluthge replied:
@christopher-dG Is this correct?
```yaml
- run: |
  julia --color=yes --check-bounds=yes --inline=${{ inputs.inline }} --project -e '
using Pkg;
kwargs = Dict{Symbol, Any}();
kwargs[:coverage] = ${{ inputs.coverage }; ${{ inputs.force_latest_compat }} && (kwargs[:force_latest_compat] = ${{ inputs.force_latest_compat }});
Pkg.test(; kwargs...)'
  shell: bash
```

christopher-dG replied:
That looks good. I think you don't even need the semicolons.

christopher-dG replied:
Except do indent the string by two more spaces.

DilumAluthge replied:
Alright, how about this:
```yaml
- run: |
  julia --color=yes --check-bounds=yes --inline=${{ inputs.inline }} --project -e '
    using Pkg
    kwargs = Dict{Symbol, Any}()
    kwargs[:coverage] = ${{ inputs.coverage }
    ${{ inputs.force_latest_compat }} && (kwargs[:force_latest_compat] = ${{ inputs.force_latest_compat }})
    Pkg.test(; kwargs...)'
  shell: bash
```

christopher-dG replied:
The whole thing needs to be indented past the parent `run`, so:

```yml
- run: |
    julia ... -e '
      using Pkg ...'
```

DilumAluthge replied:
Like this?

```yaml
- run: |
    julia --color=yes --check-bounds=yes --inline=${{ inputs.inline }} --project -e '
      using Pkg
      kwargs = Dict{Symbol, Any}()
      kwargs[:coverage] = ${{ inputs.coverage }
      ${{ inputs.force_latest_compat }} && (kwargs[:force_latest_compat] = ${{ inputs.force_latest_compat }})
      Pkg.test(; kwargs...)'
  shell: bash
```

christopher-dG replied:
Yep that parses. This seems like the dumbest nitpick ever but YAML really is this picky 😅

DilumAluthge replied:
Third time's the charm 😂 

Yeah this YAML stuff always kills me.

I've updated this PR - can you double check the syntax one last time?

Also, I've updated https://github.com/julia-actions/composite-runtest/pull/1 with the same change - can you check the syntax there as well?

christopher-dG replied:
Both look good to me.

christopher-dG commented:
Maybe there should be a version check if the feature is turned on? If we're less than 1.7, print a warning or something.

DilumAluthge commented:
> Maybe there should be a version check if the feature is turned on? If we're less than 1.7, print a warning or something.

Good idea! Take a look at the commit I just pushed - does that meet the requirement?

DilumAluthge commented:
If it looks good, I'll add it to the other PR as well.

SaschaMann commented on action.yml:
```diff
@@ -33,5 +36,17 @@ runs:
+        julia --color=yes --check-bounds=yes --inline=${{ inputs.inline }} --project -e '
+          using Pkg
+          kwargs = Dict{Symbol, Any}()
+          kwargs[:coverage] = ${{ inputs.coverage }
```
```suggestion
          kwargs[:coverage] = ${{ inputs.coverage }}
```

SaschaMann commented on action.yml:
```diff
@@ -33,5 +36,17 @@ runs:
+          kwargs = Dict{Symbol, Any}()
+          kwargs[:coverage] = ${{ inputs.coverage }
+          if ${{ inputs.force_latest_compat }}
+              if VERSION > VersionNumber(1, 6, typemax(UInt32)) # excludes 1.6, includes 1.7-DEV, includes 1.7
```
```suggestion
              if VERSION >= v"1.7.0-" # excludes 1.6, includes 1.7-DEV, includes 1.7
```

I think this is equivalent but slightly easier to read:

```julia
julia> v = v"1.7.0-"
v"1.7.0-"

julia> v"1.6.4294967295" >= v
false

julia> v"1.7-DEV" >= v
true

julia> v"1.7.0-" >= v
true

julia> v"1.7.0-DEV" >= v
true

julia> v"1.7.0-alpha" >= v
true

julia> v"1.7.0-rc1" >= v
true

julia> v"1.7.0-0" >= v
true
```

SaschaMann commented:
I wonder if instead of adding individual inputs for `Pkg.test(;kwargs...)` we should take a single input for all kwargs? Or add inputs for every kwarg?

DilumAluthge commented:
> I wonder if instead of adding individual inputs for `Pkg.test(;kwargs...)` we should take a single input for all kwargs? Or add inputs for every kwarg?

I don't feel strongly either way. What do you think users would prefer?

DilumAluthge commented:
> I wonder if instead of adding individual inputs for `Pkg.test(;kwargs...)` we should take a single input for all kwargs? Or add inputs for every kwarg?

So, we've already added individual inputs for:
- `coverage`
- `depwarn`
- `inline`

So I think probably it would be easiest to just keep the convention of having individual inputs?

SaschaMann commented:
Sounds good. 

Changes look good to me, I'll merge when the JuliaLang/Pkg.jl#2176 is merged (and leave it on change request until then to prevent accidental merges)

christopher-dG commented on action.yml:
```diff
@@ -36,5 +38,18 @@ runs:
+              if VERSION >= v"1.7.0-" # excludes 1.6, includes 1.7-DEV, includes 1.7
+                  kwargs[:force_latest_compat] = true
+              else
+                  @warn("The `force_latest_compat` option requires at least Julia 1.7 or greater", VERSION)
```
```suggestion
                  @warn("The `force_latest_compat` option requires at least Julia 1.7", VERSION)
```

DilumAluthge commented:
I've added the "auto-detect Dependabot/CompatHelper" functionality from https://github.com/JuliaLang/Pkg.jl/pull/2234

@christopher-dG @SaschaMann Can you review again?

SaschaMann commented on action.yml:
```diff
@@ -7,14 +7,16 @@ branding:
   depwarn:
     description: 'Value passed to the --depwarn flag. Options: yes | no | error. Default value: yes.'
+  force_latest_compat:
+    description: 'If true, then, for each `[compat]` entry in the active project, only the latest compatible version will be allowed. Options: true | false | autodetect. Default value: autodetect. If the value is autodetect, then force_latest_compat will be set to true if this is a Dependabot/CompatHelper pull request, and false otherwise.'
+    default: 'autodetect'
```
```suggestion
    description: 'If true, then, for each `[compat]` entry in the active project, only the latest compatible version will be allowed. Options: true | false | autodetect. Default value: auto. If the value is auto, then force_latest_compat will be set to true if this is a Dependabot/CompatHelper pull request, and false otherwise.'
    default: 'auto'
```

SaschaMann commented on autodetect-dependabot.jl:
```diff
@@ -0,0 +1,33 @@
+
+function chop_refs_head(branch_name::AbstractString)
+    if startswith(branch_name, "refs/heads/")
+        return chop(branch_name; head = 11, tail = 0)
```
Could this be done with `replace` instead? I feel like that might be nicer than hard-coding the length.

DilumAluthge replied:
Done.

christopher-dG commented on autodetect-dependabot.jl:
```diff
@@ -0,0 +1,33 @@
+module AutodetectDependabot
+
+function get_possible_branch_names()
```
None of this is necessary if it's only for GHA.

DilumAluthge replied:
Good catch. I'll fix that now.

DilumAluthge replied:
Alright I've removed the Travis and AppVeyor stuff.

SaschaMann commented on autodetect-dependabot.jl:
```diff
@@ -0,0 +1,33 @@
+end
+
+function is_dependabot_branch(branch_name::AbstractString)
+    return startswith(branch_name, "dependabot/") || startswith(branch_name, "compathelper/")
```
I think we might want this to be `dependabot/julia` (guessing based on https://github.com/julia-actions/setup-julia/pull/47), so that it doesn't affect non-Julia Dependabot PRs in mixed repos.

DilumAluthge replied:
@christopher-dG Can we ensure that Julia Dependabot PRs start with `dependabot/julia`?

christopher-dG replied:
Yes

DilumAluthge replied:
Done!

SaschaMann commented on action.yml:
```diff
@@ -7,14 +7,16 @@ branding:
   depwarn:
     description: 'Value passed to the --depwarn flag. Options: yes | no | error. Default value: yes.'
+  force_latest_compat:
+    description: 'If true, then, for each `[compat]` entry in the active project, only the latest compatible version will be allowed. Options: true | false | autodetect. Default value: auto. If the value is auto, then force_latest_compat will be set to true if this is a Dependabot/CompatHelper pull request, and false otherwise.'
```
```suggestion
    description: 'If true, then, for each `[compat]` entry in the active project, only the latest compatible version will be allowed. Options: true | false | auto. Default value: auto. If the value is auto, then force_latest_compat will be set to true if this is a Dependabot/CompatHelper pull request, and false otherwise.'
```

Sorry, I missed one.

SaschaMann commented:
Should this be considered a breaking change and require a major version bump of the action?

DilumAluthge commented:
Hmmm. The only change in behavior is on CompatHelper/Dependabot PRs, right?

Dependabot doesn't exist yet.

And for CompatHelper, this was always supposed to be the intended behavior, so honestly I would consider it to be a bug fix.

SaschaMann commented on autodetect-dependabot.jl:
```diff
@@ -0,0 +1,29 @@
+
+function chop_refs_head(branch_name::AbstractString)
+    if startswith(branch_name, "refs/heads/")
+        return replace(branch_name, "refs/heads/" => "")
```
For the (very unlikely) case that someone names their branch `refs/heads`, let's only replace the first occurence

```suggestion
        return replace(branch_name, "refs/heads/" => "", count = 1)
```

SaschaMann replied:
For the (very unlikely) case that someone names their branch `refs/heads`, let's only replace the first occurence and only if it's at the start of the string:

```suggestion
        return replace(branch_name, r"^(refs\/heads\/)" => "")
```

DilumAluthge commented:
As long as the behavior change is only on Dependabot/CompatHelper PRs, I would say this "technically is a breaking change in behavior, but really more of a bug fix, so it's okay to do it in a patch release".

DilumAluthge commented:
> the behavior change is only on Dependabot/CompatHelper PRs

As long as this is true.

DilumAluthge commented:
I've added another small change: if the user sets `force_latest_compat` to `auto` (the default value), and this is in fact a Dependabot/CompatHelper job, then I print an `@info` message saying:
> @info("This is a Dependabot/CompatHelper job, so `force_latest_compat` has been set to `true`")

That way, if a user looks at the logs, they'll know that `force_latest_compat` was set to be `true`.

SaschaMann commented:
Looks good to me :)

DilumAluthge commented:
Once https://github.com/JuliaLang/Pkg.jl/pull/2439 is merged, I'll fix the merge conflicts and rebase this PR on the latest master.

DilumAluthge commented:
@christopher-dG @SaschaMann We are still waiting on https://github.com/JuliaLang/Pkg.jl/pull/2439 to be merged, but in the mean time, I have rebased and fixed the merge conflicts. Can you review this again?

DilumAluthge commented:
We'll also need to wait until:
1. https://github.com/JuliaLang/julia/pull/40266 is merged into master
2. The nightly binaries are built and unloaded from that commit on master

DilumAluthge commented:
Okay, https://github.com/JuliaLang/julia/pull/40266 has been merged.

DilumAluthge commented:
FYI the macOS Buildbots are currently broken.

SaschaMann commented on action.yml:
```diff
@@ -13,6 +13,9 @@ inputs:
     description: 'Value passed to the --depwarn flag. Options: yes | no | error. Default value: yes.'
     default: 'yes'
+  force_latest_compatible_version:
+    description: 'If true, then, for each `[compat]` entry in the active project, only the latest compatible version will be allowed. Options: true | false | auto. Default value: auto. If the value is auto, then force_latest_compatible_version will be set to true if this is a Dependabot/CompatHelper pull request, and false otherwise.'
```
```suggestion
    description: 'If true, then, for each [compat] entry in the active project, only allow the latest compatible version. If the value is auto and the pull request has been opened by Dependabot or CompatHelper, then force_latest_compatible_version will be set to true, otherwise it will be set to false. Options: true | false | auto. Default value: auto.'
```

SaschaMann commented on action.yml:
```diff
@@ -44,13 +47,18 @@ runs:
-
-        # Run the Julia command
-        "${julia_cmd[@]}"
+        JULIA_CMD="julia"
+        PREFIX="${{ inputs.prefix }}"
+        [[ -n $PREFIX ]] && JULIA_CMD="$PREFIX $JULIA_CMD"
+        echo "JULIA_CMD=$JULIA_CMD" >> $GITHUB_ENV
+      shell: bash
+    - run: |
+        ${{ env.JULIA_CMD }} --check-bounds=yes --color=yes --depwarn=${{ inputs.depwarn }} --inline=${{ inputs.inline }} --project=${{ inputs.project }} -e '
+          import Pkg
+          include(joinpath(ENV["GITHUB_ACTION_PATH"], "kwargs.jl"))
+          kwargs = Kwargs.kwargs(;
+            coverage = :(${{ inputs.coverage }}),
+            force_latest_compatible_version = :(${{ inputs.force_latest_compatible_version }}),
+          )
+          Pkg.test(; kwargs...)'
```
Why is this split into two steps?

DilumAluthge replied:
Would I just change `${{ env.JULIA_CMD }}` to `"$JULIA_CMD"`

SaschaMann replied:
You could leave the bash part almost as it was and then only replace `-e 'using Pkg; Pkg.test(coverage=${{ inputs.coverage }})'` with the new code.

I forgot the exact details, but there was some reason why `()` and a bash array was used. Originally I had a similar approach as this with normal double quoted variables.

SaschaMann replied:
Other than that, this PR is good to be merged

DilumAluthge replied:
@SaschaMann Take a look now. Is this correct?

SaschaMann replied:
Looks good to me! Thanks :)

SaschaMann commented on kwargs.jl:
```diff
@@ -0,0 +1,37 @@
+        if force_latest_compatible_version isa Bool
+            kwargs_dict[:force_latest_compatible_version] = force_latest_compatible_version
+        elseif force_latest_compatible_version == :auto
+            if is_dependabot_job
+                @info("This is a Dependabot/CompatHelper job, so `force_latest_compatible_version` has been set to `true`")
+                kwargs_dict[:force_latest_compatible_version] = true
+            else
+                kwargs_dict[:force_latest_compatible_version] = false
+            end
```
```suggestion
            is_dependabot_job && @info("This is a Dependabot/CompatHelper job, so `force_latest_compatible_version` has been set to `true`")
            kwargs_dict[:force_latest_compatible_version] = is_dependabot_job
```

(Untested and written in the GH suggestion prompt)

Remove another nesting level here will also help make understanding what's going on easier, I think.

SaschaMann commented on kwargs.jl:
```diff
@@ -0,0 +1,37 @@
+                kwargs_dict[:force_latest_compatible_version] = false
+            end
+        else
+            throw(ArgumentError("Invalid value for force_latest_compatible_version: $(force_latest_compatible_version)"))
```
Could you move this to a guard clause all the way at the top of the function, please?

DilumAluthge replied:
Done

SaschaMann commented on autodetect-dependabot.jl:
```diff
@@ -0,0 +1,29 @@
+end
+
+function _chop_refs_head(branch_name::AbstractString)
+    if startswith(branch_name, "refs/heads/")
+        return replace(branch_name, r"^(refs\/heads\/)" => "")
+    else
+        return branch_name
+    end
```
```suggestion
    replace(branch_name, r"^(refs\/heads\/)" => "")
```

If the regex doesn't match anything, nothing will be replaced. We don't need to check for that case manually.

SaschaMann reviewed (commented):
Thanks for keeping this up-to-date!

I found it a bit hard to follow what's going on in the script. I left a few comments on how, imo, that could be improved, mostly by removing nesting levels.

Happy to apply these changes myself in a follow-up PR too

DilumAluthge commented on kwargs.jl:
```diff
@@ -0,0 +1,37 @@
+    kwargs_dict = Dict{Symbol, Any}()
+    kwargs_dict[:coverage] = coverage
+
+    if !Sys.isapple() && VERSION >= v"1.7.0-" # TODO: delete this line
```
macOS buildbots have been broken for approximately a week, so the nightlies  for macOS are out of date. A PR was just merged that will fix this, so soon the macOS nightlies will be up-to-date, and we can remove this before merging this PR.

DilumAluthge commented on kwargs.jl:
```diff
@@ -0,0 +1,37 @@
+            )
+        end
+    end
+    return kwargs_dict
```
Done

DilumAluthge commented:
@SaschaMann Take a look now.

DilumAluthge commented:
We'll need to wait to merge until the macOS nightlies are fixed.

DilumAluthge commented:
Alright the macOS buildbots are fixed and the macOS nightlies are up to date.

This is good to merge from my end.

@SaschaMann

DilumAluthge commented:
@SaschaMann Could we tag a new v1.x.y tag?

DilumAluthge commented:
You beat me to it 😂 

https://github.com/julia-actions/julia-runtest/releases/tag/v1.6.0

DilumAluthge commented:
@SaschaMann and @christopher-dG thank you for your help on this?

SaschaMann commented:
I probably should have mentioned that I'm gonna tag it :D 

I usually do that right after merging PRs to these actions, I should probably setup a merge-and-tag script at some point.

DilumAluthge commented:
🚀 🚀 🚀 it works!!! 

https://github.com/cscherrer/SossMLJ.jl/runs/2241088439 (triggered by https://github.com/cscherrer/SossMLJ.jl/pull/132)
```
┌ Error: Package is not at the latest compatible version
│   name = "CategoricalArrays"
│   uuid = UUID("324d7699-5711-5eae-9e2f-1d82baa6b597")
│   compat_entry_string = "0.8, 0.9"
│   active_version = v"0.8.3"
│   latest_compatible_version = v"0.9.4"
│   earliest_backwards_compatible_version = v"0.9.0"
│   allow_earlier_backwards_compatible_versions = true
└ @ Pkg.Operations /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.7/Pkg/src/Operations.jl:1856
ERROR: One or more direct dependencies is not at the latest compatible version
```

DilumAluthge commented:
> I should probably setup a merge-and-tag script at some point.

This could be useful!
